CI: Integration tests - Test_meshEndpointSlice_Reconcile

## CI failure

```
/tmp/go-stress-20240606T093130-3062731158
I0606 09:32:49.496892  156987 endpointslice_controller.go:296] "Starting endpoint slice controller"
I0606 09:32:49.496976  156987 shared_informer.go:313] Waiting for caches to sync for endpoint_slice
I0606 09:32:49.497011  156987 shared_informer.go:320] Caches are synced for endpoint_slice
--- FAIL: Test_meshEndpointSlice_Reconcile (0.31s)
    logger.go:257: time=2024-06-06T09:32:49.294+02:00 level=INFO source=/home/marco/Documents/isovalent/cilium/vendor/github.com/cilium/hive/hive.go:337 msg=Starting
    logger.go:257: time=2024-06-06T09:32:49.294+02:00 level=INFO source=/home/marco/Documents/isovalent/cilium/vendor/github.com/cilium/hive/hive.go:341 msg=Started duration=95.581µs
    --- FAIL: Test_meshEndpointSlice_Reconcile/Create_global_service_then_service (0.01s)
        endpointslice_test.go:189: 
                Error Trace:    /home/marco/Documents/isovalent/cilium/pkg/clustermesh/endpointslicesync/endpointslice_test.go:189
                Error:          Not equal: 
                                expected: 1
                                actual  : 0
                Test:           Test_meshEndpointSlice_Reconcile/Create_global_service_then_service
    logger.go:257: time=2024-06-06T09:32:49.590+02:00 level=INFO source=/home/marco/Documents/isovalent/cilium/vendor/github.com/cilium/hive/hive.go:353 msg=Stopping
    logger.go:257: time=2024-06-06T09:32:49.590+02:00 level=INFO source=/home/marco/Documents/isovalent/cilium/pkg/hive/health/provider.go:62 msg="health.job-module-status-metrics (rev=2)" module=health
FAIL
```

I managed to reproduce it locally (although it seems fairly rare) with:

```
$ go install golang.org/x/tools/cmd/stress@latest
$ go test -c -race ./pkg/clustermesh/endpointslicesync
$ stress -p 250 ./endpointslicesync.test
```

/cc @MrFreezeex

--- a/pkg/clustermesh/common/services.go
+++ b/pkg/clustermesh/common/services.go
@@ -8,7 +8,6 @@ import (
 
 	"github.com/sirupsen/logrus"
 	"k8s.io/apimachinery/pkg/types"
-	"k8s.io/apimachinery/pkg/util/sets"
 
 	"github.com/cilium/cilium/pkg/kvstore/store"
 	"github.com/cilium/cilium/pkg/lock"
@@ -28,9 +27,8 @@ func newGlobalService() *GlobalService {
 }
 
 type GlobalServiceCache struct {
-	mutex       lock.RWMutex
-	byName      map[types.NamespacedName]*GlobalService
-	byNamespace map[string]sets.Set[*GlobalService]
+	mutex  lock.RWMutex
+	byName map[types.NamespacedName]*GlobalService
 
 	// metricTotalGlobalServices is the gauge metric for total of global services
 	metricTotalGlobalServices metric.Gauge
@@ -39,7 +37,6 @@ type GlobalServiceCache struct {
 func NewGlobalServiceCache(metricTotalGlobalServices metric.Gauge) *GlobalServiceCache {
 	return &GlobalServiceCache{
 		byName:                    map[types.NamespacedName]*GlobalService{},
-		byNamespace:               map[string]sets.Set[*GlobalService]{},
 		metricTotalGlobalServices: metricTotalGlobalServices,
 	}
 }
@@ -89,25 +86,6 @@ func (c *GlobalServiceCache) GetGlobalService(serviceNN types.NamespacedName) *G
 	return nil
 }
 
-// GetServices returns the services for a specific namespace. This function does not
-// make copy of the cluster services objects so those objects should not be mutated.
-func (c *GlobalServiceCache) GetServices(namespace string) []*serviceStore.ClusterService {
-	c.mutex.RLock()
-	defer c.mutex.RUnlock()
-
-	clusterSvcs := []*serviceStore.ClusterService{}
-
-	if globalSvcs, ok := c.byNamespace[namespace]; ok {
-		for globalSvc := range globalSvcs {
-			for _, clusterSvc := range globalSvc.ClusterServices {
-				clusterSvcs = append(clusterSvcs, clusterSvc)
-			}
-		}
-	}
-
-	return clusterSvcs
-}
-
 func (c *GlobalServiceCache) OnUpdate(svc *serviceStore.ClusterService) {
 	scopedLog := log.WithFields(logrus.Fields{
 		logfields.ServiceName: svc.String(),
@@ -123,12 +101,6 @@ func (c *GlobalServiceCache) OnUpdate(svc *serviceStore.ClusterService) {
 		c.byName[svc.NamespaceServiceName()] = globalSvc
 		scopedLog.Debugf("Created global service %s", svc.NamespaceServiceName())
 		c.metricTotalGlobalServices.Set(float64(len(c.byName)))
-		globalSvcs, ok := c.byNamespace[svc.Namespace]
-		if !ok {
-			globalSvcs = sets.Set[*GlobalService]{}
-			c.byNamespace[svc.Namespace] = globalSvcs
-		}
-		globalSvcs.Insert(globalSvc)
 	}
 
 	scopedLog.Debugf("Updated service definition of remote cluster %#v", svc)
@@ -155,10 +127,6 @@ func (c *GlobalServiceCache) delete(globalService *GlobalService, clusterName st
 	// After the last cluster service is removed, remove the global service
 	if len(globalService.ClusterServices) == 0 {
 		scopedLog.Debugf("Deleted global service %s", serviceNN.String())
-		c.byNamespace[serviceNN.Namespace].Delete(globalService)
-		if len(c.byNamespace[serviceNN.Namespace]) == 0 {
-			delete(c.byNamespace, serviceNN.Namespace)
-		}
 		delete(c.byName, serviceNN)
 		c.metricTotalGlobalServices.Set(float64(len(c.byName)))
 	}

--- a/pkg/clustermesh/endpointslicesync/endpointslice_meta_controller.go
+++ b/pkg/clustermesh/endpointslicesync/endpointslice_meta_controller.go
@@ -64,7 +64,9 @@ func newEndpointSliceMeshController(
 	factory := informers.NewSharedInformerFactory(meshClient, 12*time.Hour)
 	endpointSliceInformer := factory.Discovery().V1().EndpointSlices()
 
-	meshServiceInformer := newMeshServiceInformer(logger, globalServices, services)
+	meshServiceInformer := newMeshServiceInformer(
+		logger, globalServices, services, meshNodeInformer,
+	)
 
 	controller := endpointslice.NewControllerWithName(
 		ctx, meshPodInformer, meshServiceInformer,

--- a/pkg/clustermesh/endpointslicesync/endpointslice_test.go
+++ b/pkg/clustermesh/endpointslicesync/endpointslice_test.go
@@ -5,22 +5,18 @@ package endpointslicesync
 
 import (
 	"context"
-	"fmt"
-	"reflect"
 	"testing"
 	"time"
-	"unsafe"
 
-	"github.com/cilium/endpointslice-controller/endpointslice"
 	"github.com/cilium/hive/cell"
 	"github.com/cilium/hive/hivetest"
 	"github.com/sirupsen/logrus"
+	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/require"
 	corev1 "k8s.io/api/core/v1"
 	discovery "k8s.io/api/discovery/v1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	cache "k8s.io/client-go/tools/cache"
-	"k8s.io/client-go/util/workqueue"
 	mcsapiv1alpha1 "sigs.k8s.io/mcs-api/pkg/apis/v1alpha1"
 
 	"github.com/cilium/cilium/operator/k8s"
@@ -41,23 +37,6 @@ const (
 	remoteClusterName string = "cluster-1"
 	globalSvcIP       string = "42.42.42.42"
 )
-
-func getControllerQueue(controller *endpointslice.Controller) workqueue.RateLimitingInterface {
-	return *(*workqueue.RateLimitingInterface)(unsafe.Pointer(
-		reflect.ValueOf(controller).Elem().FieldByName("queue").UnsafeAddr(),
-	))
-}
-
-func waitEmptyQueue(queue workqueue.RateLimitingInterface) error {
-	for i := 0; i < 20; i++ {
-		time.Sleep(time.Millisecond * 10)
-		if queue.Len() == 0 {
-			return nil
-		}
-	}
-
-	return fmt.Errorf("endpointSlice controller queue has not been emptied after 200ms")
-}
 
 func createService(name string) *slim_corev1.Service {
 	return &slim_corev1.Service{
@@ -80,7 +59,7 @@ func createGlobalService(
 	globalService *common.GlobalServiceCache,
 	podInformer *meshPodInformer,
 	svcName string,
-	updateClusterSvc func(*store.ClusterService)) {
+	updateClusterSvc func(*store.ClusterService)) *store.ClusterService {
 	clusterSvc := &store.ClusterService{
 		Cluster:   remoteClusterName,
 		Namespace: svcName,
@@ -97,7 +76,7 @@ func createGlobalService(
 	globalService.OnUpdate(clusterSvc)
 	// We manually call the rest of the informer for convenience
 	podInformer.onClusterServiceUpdate(clusterSvc)
-
+	return clusterSvc
 }
 
 func getEndpointSlice(clientset k8sClient.Clientset, svcName string) (*discovery.EndpointSliceList, error) {
@@ -149,6 +128,9 @@ func Test_meshEndpointSlice_Reconcile(t *testing.T) {
 
 	svcStore, _ := services.Store(context.Background())
 
+	tick := 10 * time.Millisecond
+	timeout := 200 * time.Millisecond
+
 	t.Run("Create service then global service", func(t *testing.T) {
 		svcName := "local-svc-global-svc"
 		hostname := svcName + "-0"
@@ -159,22 +141,28 @@ func Test_meshEndpointSlice_Reconcile(t *testing.T) {
 			clusterSvc.Hostnames = map[string]string{globalSvcIP: hostname}
 		})
 
-		queue := getControllerQueue(controller)
-		require.NoError(t, waitEmptyQueue(queue))
+		var epList *discovery.EndpointSliceList
+		require.EventuallyWithT(t, func(c *assert.CollectT) {
+			epList, err = getEndpointSlice(&fakeClient, svcName)
+			assert.NoError(c, err)
+			assert.Len(c, epList.Items, 1)
+		}, timeout, tick, "endpointslice is not reconciled correctly")
 
-		epList, err := getEndpointSlice(&fakeClient, svcName)
-		require.NoError(t, err)
-		require.Equal(t, 1, len(epList.Items))
 		require.Equal(t, map[string]string{
 			discovery.LabelServiceName:        svcName,
 			discovery.LabelManagedBy:          utils.EndpointSliceMeshControllerName,
 			mcsapiv1alpha1.LabelSourceCluster: remoteClusterName,
 			corev1.IsHeadlessService:          "",
 		}, epList.Items[0].Labels)
-		require.Equal(t, 1, len(epList.Items[0].Endpoints))
+		require.Len(t, epList.Items[0].Endpoints, 1)
 
-		require.NotNil(t, 1, epList.Items[0].Endpoints[0].Hostname)
-		require.Equal(t, *epList.Items[0].Endpoints[0].Hostname, hostname)
+		require.NotNil(t, epList.Items[0].Endpoints[0].Hostname)
+		require.Equal(t, hostname, *epList.Items[0].Endpoints[0].Hostname)
+
+		require.Len(t, epList.Items[0].OwnerReferences, 1)
+		require.Equal(t, "v1", epList.Items[0].OwnerReferences[0].APIVersion)
+		require.Equal(t, "Service", epList.Items[0].OwnerReferences[0].Kind)
+		require.Equal(t, "local-svc-global-svc", epList.Items[0].OwnerReferences[0].Name)
 	})
 
 	t.Run("Create global service then service", func(t *testing.T) {
@@ -184,44 +172,19 @@ func Test_meshEndpointSlice_Reconcile(t *testing.T) {
 		svcStore.CacheStore().Add(svc1)
 		serviceInformer.refreshAllCluster(svc1)
 
-		queue := getControllerQueue(controller)
-		require.NoError(t, waitEmptyQueue(queue))
+		var epList *discovery.EndpointSliceList
+		require.EventuallyWithT(t, func(c *assert.CollectT) {
+			epList, err = getEndpointSlice(&fakeClient, svcName)
+			assert.NoError(t, err)
+			assert.Len(t, epList.Items, 1)
+		}, timeout, tick, "endpointslice is not reconciled correctly")
 
-		epList, err := getEndpointSlice(&fakeClient, svcName)
-		require.NoError(t, err)
-		require.Equal(t, 1, len(epList.Items))
 		require.Equal(t, map[string]string{
 			discovery.LabelServiceName:        svcName,
 			discovery.LabelManagedBy:          utils.EndpointSliceMeshControllerName,
 			mcsapiv1alpha1.LabelSourceCluster: remoteClusterName,
 			corev1.IsHeadlessService:          "",
 		}, epList.Items[0].Labels)
-	})
-
-	t.Run("Create service without global service", func(t *testing.T) {
-		svcName := "local-svc-no-global-svc"
-		svc1 := createService(svcName)
-		svcStore.CacheStore().Add(svc1)
-		serviceInformer.refreshAllCluster(svc1)
-
-		queue := getControllerQueue(controller)
-		require.NoError(t, waitEmptyQueue(queue))
-
-		epList, err := getEndpointSlice(&fakeClient, svcName)
-		require.NoError(t, err)
-		require.Zero(t, len(epList.Items))
-	})
-
-	t.Run("Create global service without service", func(t *testing.T) {
-		svcName := "global-svc-no-local-svc"
-		createGlobalService(globalService, podInformer, svcName, nil)
-
-		queue := getControllerQueue(controller)
-		require.NoError(t, waitEmptyQueue(queue))
-
-		epList, err := getEndpointSlice(&fakeClient, svcName)
-		require.NoError(t, err)
-		require.Zero(t, len(epList.Items))
 	})
 
 	t.Run("Create headless service and global service", func(t *testing.T) {
@@ -233,29 +196,11 @@ func Test_meshEndpointSlice_Reconcile(t *testing.T) {
 		serviceInformer.refreshAllCluster(svc1)
 		createGlobalService(globalService, podInformer, svcName, nil)
 
-		queue := getControllerQueue(controller)
-		require.NoError(t, waitEmptyQueue(queue))
-
-		epList, err := getEndpointSlice(&fakeClient, svcName)
-		require.NoError(t, err)
-		require.Equal(t, 1, len(epList.Items))
-	})
-
-	t.Run("Create headless service with endpointslice opt-out and global service", func(t *testing.T) {
-		svcName := "local-svc-headless-opt-out-global-svc"
-		svc1 := createService(svcName)
-		svc1.Spec.ClusterIP = corev1.ClusterIPNone
-		svc1.ObjectMeta.Annotations[annotation.GlobalServiceSyncEndpointSlices] = "false"
-		svcStore.CacheStore().Add(svc1)
-		serviceInformer.refreshAllCluster(svc1)
-		createGlobalService(globalService, podInformer, svcName, nil)
-
-		queue := getControllerQueue(controller)
-		require.NoError(t, waitEmptyQueue(queue))
-
-		epList, err := getEndpointSlice(&fakeClient, svcName)
-		require.NoError(t, err)
-		require.Equal(t, 0, len(epList.Items))
+		require.EventuallyWithT(t, func(c *assert.CollectT) {
+			epList, err := getEndpointSlice(&fakeClient, svcName)
+			assert.NoError(c, err)
+			assert.Len(c, epList.Items, 1)
+		}, timeout, tick, "endpointslice is not reconciled correctly")
 	})
 
 	t.Run("Create service with global annotation and remove it", func(t *testing.T) {
@@ -265,22 +210,23 @@ func Test_meshEndpointSlice_Reconcile(t *testing.T) {
 		svcStore.CacheStore().Add(svc1)
 		serviceInformer.refreshAllCluster(svc1)
 
-		queue := getControllerQueue(controller)
-		require.NoError(t, waitEmptyQueue(queue))
-
-		// Make sure that we have 1 endpointslice
-		epList, err := getEndpointSlice(&fakeClient, svcName)
-		require.NoError(t, err)
-		require.Equal(t, 1, len(epList.Items))
+		var epList *discovery.EndpointSliceList
+		require.EventuallyWithT(t, func(c *assert.CollectT) {
+			// Make sure that we have 1 endpointslice
+			epList, err = getEndpointSlice(&fakeClient, svcName)
+			assert.NoError(c, err)
+			assert.Len(c, epList.Items, 1)
+		}, timeout, tick, "endpointslice is not reconciled correctly")
 
 		svc1.Annotations[annotation.GlobalServiceSyncEndpointSlices] = "false"
 		svcStore.CacheStore().Update(svc1)
 		serviceInformer.refreshAllCluster(svc1)
 
-		require.NoError(t, waitEmptyQueue(queue))
+		require.EventuallyWithT(t, func(c *assert.CollectT) {
+			epList, err := getEndpointSlice(&fakeClient, svcName)
+			assert.NoError(c, err)
+			assert.Len(c, epList.Items, 0)
+		}, timeout, tick, "endpointslice is not reconciled correctly")
 
-		epList, err = getEndpointSlice(&fakeClient, svcName)
-		require.NoError(t, err)
-		require.Zero(t, len(epList.Items))
 	})
 }

--- a/pkg/clustermesh/endpointslicesync/endpointslice_test.go
+++ b/pkg/clustermesh/endpointslicesync/endpointslice_test.go
@@ -227,6 +227,31 @@ func Test_meshEndpointSlice_Reconcile(t *testing.T) {
 			assert.NoError(c, err)
 			assert.Len(c, epList.Items, 0)
 		}, timeout, tick, "endpointslice is not reconciled correctly")
+	})
 
+	t.Run("Create service and global service and then delete global svc", func(t *testing.T) {
+		svcName := "local-svc-no-global-svc"
+		clusterSvc := createGlobalService(globalService, podInformer, svcName, nil)
+		svc1 := createService(svcName)
+		svcStore.CacheStore().Add(svc1)
+		serviceInformer.refreshAllCluster(svc1)
+
+		var epList *discovery.EndpointSliceList
+		require.EventuallyWithT(t, func(c *assert.CollectT) {
+			// Make sure that we have 1 endpointslice
+			epList, err = getEndpointSlice(&fakeClient, svcName)
+			assert.NoError(c, err)
+			assert.Len(c, epList.Items, 1)
+		}, timeout, tick, "endpointslice is not reconciled correctly")
+
+		globalService.OnDelete(clusterSvc)
+		// We manually call the rest of the informer for convenience
+		podInformer.onClusterServiceDelete(clusterSvc)
+
+		require.EventuallyWithT(t, func(c *assert.CollectT) {
+			epList, err := getEndpointSlice(&fakeClient, svcName)
+			assert.NoError(c, err)
+			assert.Len(c, epList.Items, 0)
+		}, timeout, tick, "endpointslice is not reconciled correctly")
 	})
 }

--- a/pkg/clustermesh/endpointslicesync/node_informer.go
+++ b/pkg/clustermesh/endpointslicesync/node_informer.go
@@ -56,6 +56,13 @@ func createDummyNode(cluster string) *v1.Node {
 	}
 }
 
+func (i *meshNodeInformer) ListClusters() []string {
+	i.mutex.RLock()
+	defer i.mutex.RUnlock()
+
+	return maps.Keys(i.nodes)
+}
+
 func (i *meshNodeInformer) List(selector labels.Selector) ([]*v1.Node, error) {
 	reqs, _ := selector.Requirements()
 	if !selector.Empty() {

--- a/pkg/clustermesh/endpointslicesync/service_informer.go
+++ b/pkg/clustermesh/endpointslicesync/service_informer.go
@@ -48,6 +48,7 @@ type meshServiceInformer struct {
 	globalServiceCache *common.GlobalServiceCache
 	services           resource.Resource[*slim_corev1.Service]
 	serviceStore       resource.Store[*slim_corev1.Service]
+	meshNodeInformer   *meshNodeInformer
 
 	servicesSynced atomic.Bool
 	handler        cache.ResourceEventHandler
@@ -99,12 +100,14 @@ func newMeshServiceInformer(
 	logger logrus.FieldLogger,
 	globalServiceCache *common.GlobalServiceCache,
 	services resource.Resource[*slim_corev1.Service],
+	meshNodeInformer *meshNodeInformer,
 ) *meshServiceInformer {
 	return &meshServiceInformer{
 		dummyInformer:      dummyInformer{name: "meshServiceInformer", logger: logger},
 		logger:             logger,
 		globalServiceCache: globalServiceCache,
 		services:           services,
+		meshNodeInformer:   meshNodeInformer,
 	}
 }
 
@@ -195,17 +198,36 @@ type meshServiceLister struct {
 	namespace string
 }
 
+// List returns the matrix of all the local services and all the remote clusters.
+// This is not similar to what does the Get method. For instance, List may returns
+// services that could not be found on a call to the Get method.
+// By doing that we can ensure that the controller reconciliation is called
+// on every possible remote services especially the one that are deleted while
+// we still have some EndpointSlices locally (which won't be cleared by the
+// OwnerReference mechanism since our actual local service is not deleted).
 func (l meshServiceLister) List(selector labels.Selector) ([]*v1.Service, error) {
 	reqs, _ := selector.Requirements()
 	if !selector.Empty() {
 		return nil, fmt.Errorf("meshServiceInformer only supports listing everything as requirements: %s", reqs)
 	}
 
-	clusterSvcs := l.informer.globalServiceCache.GetServices(l.namespace)
-	svcs := make([]*v1.Service, 0, len(clusterSvcs))
-	for _, clusterSvc := range clusterSvcs {
-		if svc, err := l.informer.clusterSvcToSvc(clusterSvc, false); err == nil {
-			svcs = append(svcs, svc)
+	svcIter := l.informer.serviceStore.IterKeys()
+	clusters := l.informer.meshNodeInformer.ListClusters()
+	var svcs []*v1.Service
+	for svcIter.Next() {
+		if svcIter.Key().Namespace != l.namespace {
+			continue
+		}
+
+		for _, cluster := range clusters {
+			dummyClusterSvc := &store.ClusterService{
+				Cluster:   cluster,
+				Name:      svcIter.Key().Name,
+				Namespace: l.namespace,
+			}
+			if svc, err := l.informer.clusterSvcToSvc(dummyClusterSvc, false); err == nil {
+				svcs = append(svcs, svc)
+			}
 		}
 	}
 

--- a/pkg/clustermesh/endpointslicesync/service_informer.go
+++ b/pkg/clustermesh/endpointslicesync/service_informer.go
@@ -23,6 +23,7 @@ import (
 
 	"github.com/cilium/cilium/pkg/annotation"
 	"github.com/cilium/cilium/pkg/clustermesh/common"
+	"github.com/cilium/cilium/pkg/k8s"
 	"github.com/cilium/cilium/pkg/k8s/resource"
 	slim_corev1 "github.com/cilium/cilium/pkg/k8s/slim/k8s/api/core/v1"
 	"github.com/cilium/cilium/pkg/service/store"
@@ -211,18 +212,18 @@ func (l meshServiceLister) List(selector labels.Selector) ([]*v1.Service, error)
 		return nil, fmt.Errorf("meshServiceInformer only supports listing everything as requirements: %s", reqs)
 	}
 
-	svcIter := l.informer.serviceStore.IterKeys()
+	originalSvcs, err := l.informer.serviceStore.ByIndex(k8s.NamespaceIndex, l.namespace)
+	if err != nil {
+		return nil, err
+	}
+
 	clusters := l.informer.meshNodeInformer.ListClusters()
 	var svcs []*v1.Service
-	for svcIter.Next() {
-		if svcIter.Key().Namespace != l.namespace {
-			continue
-		}
-
+	for _, svc := range originalSvcs {
 		for _, cluster := range clusters {
 			dummyClusterSvc := &store.ClusterService{
 				Cluster:   cluster,
-				Name:      svcIter.Key().Name,
+				Name:      svc.Name,
 				Namespace: l.namespace,
 			}
 			if svc, err := l.informer.clusterSvcToSvc(dummyClusterSvc, false); err == nil {

--- a/pkg/k8s/resource_ctors.go
+++ b/pkg/k8s/resource_ctors.go
@@ -45,10 +45,24 @@ var DefaultConfig = Config{
 	EnableK8sEndpointSlice: true,
 }
 
+const (
+	NamespaceIndex = "namespace"
+)
+
 // Flags implements the cell.Flagger interface.
 func (def Config) Flags(flags *pflag.FlagSet) {
 	flags.Bool("enable-k8s-endpoint-slice", def.EnableK8sEndpointSlice, "Enables k8s EndpointSlice feature in Cilium if the k8s cluster supports it")
 	flags.String("k8s-service-proxy-name", def.K8sServiceProxyName, "Value of K8s service-proxy-name label for which Cilium handles the services (empty = all services without service.kubernetes.io/service-proxy-name label)")
+}
+
+// namespaceIndexFunc is an IndexFunc that indexes Namespace of Kubernetes
+// types by their namespace.
+func namespaceIndexFunc(obj any) ([]string, error) {
+	object, ok := obj.(utils.NamespaceNameGetter)
+	if !ok {
+		return nil, fmt.Errorf("unexpected object type: %T", obj)
+	}
+	return []string{object.GetNamespace()}, nil
 }
 
 // ServiceResource builds the Resource[Service] object.
@@ -60,11 +74,18 @@ func ServiceResource(lc cell.Lifecycle, cfg Config, cs client.Clientset, opts ..
 	if err != nil {
 		return nil, err
 	}
+	indexers := cache.Indexers{
+		NamespaceIndex: namespaceIndexFunc,
+	}
 	lw := utils.ListerWatcherWithModifiers(
 		utils.ListerWatcherFromTyped[*slim_corev1.ServiceList](cs.Slim().CoreV1().Services("")),
 		append(opts, optsModifier)...,
 	)
-	return resource.New[*slim_corev1.Service](lc, lw, resource.WithMetric("Service")), nil
+	return resource.New[*slim_corev1.Service](
+		lc, lw,
+		resource.WithMetric("Service"),
+		resource.WithIndexers(indexers),
+	), nil
 }
 
 func NodeResource(lc cell.Lifecycle, cs client.Clientset, opts ...func(*metav1.ListOptions)) (resource.Resource[*slim_corev1.Node], error) {

